File: android/app/src/main/java/com/foreverjukebox/app/engine/Random.kt
package com.foreverjukebox.app.engine

enum class RandomMode {
    Random,
    Seeded,
    Deterministic
}

fun createRng(mode: RandomMode, seed: Int? = null): () -> Double {
    if (mode == RandomMode.Random) {
        return { kotlin.random.Random.nextDouble() }
    }
    var t = seed ?: 123456789
    return {
        t += 0x6d2b79f5
        var x = t
        x = (x xor (x ushr 15)) * (x or 1)
        x = x xor (x + (x xor (x ushr 7)) * (x or 61))
        ((x xor (x ushr 14)).toLong() and 0xffffffffL).toDouble() / 4294967296.0
    }
}

File: web/src/engine/random.ts
export type RandomMode = "random" | "seeded" | "deterministic";

export function createRng(mode: RandomMode, seed?: number): () => number {
  if (mode === "random") {
    return Math.random;
  }
  let t = seed ?? 123456789;
  return function rng() {
    t += 0x6d2b79f5;
    let x = t;
    x = Math.imul(x ^ (x >>> 15), x | 1);
    x ^= x + Math.imul(x ^ (x >>> 7), x | 61);
    return ((x ^ (x >>> 14)) >>> 0) / 4294967296;
  };
}
